Adding SHA* implementations using CryptoKit

nmahendru opened the pull request:
#### 2af6dd861a06a3881f98e2f872c49ea04c3533a9
<pre>
Adding SHA* implementations using CryptoKit
<a href="https://rdar.apple.com/122317816">rdar://122317816</a>

Reviewed by Alex Christensen.

This PR introduces the CryptoDigest implementations from CryptoKit.

This change also moves the CryptoKitShim into PAL from WebCore to avoid circular dependencies
from WebCore &lt;== PAL. The CryptoDigest Implementation resides in PAL and still needs to use the Shim.
We also add a negative test to for zero length input digests to ensure the new implementation supports it.

Another big change is that the Digest implementations have been replaced with a one shot implementation
instead of init, update and finalize style. That helps simplify the swift shim and does not mandate dealing with swift
allocated memory in C++ land.
The remnants of the init/update/finalize style have been
left in because of this(Just in case we have to implement it in the future):
<a href="https://github.com/wintercg/proposal-webcrypto-streams/blob/main/explainer.md">https://github.com/wintercg/proposal-webcrypto-streams/blob/main/explainer.md</a>

Another change is to introduce smart pointers for managing memory wherever possible.

* LayoutTests/crypto/subtle/sha1-digest.html:
* LayoutTests/crypto/subtle/sha256-digest.html:
* LayoutTests/crypto/subtle/sha384-digest.html:
* LayoutTests/crypto/subtle/sha512-digest.html:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreSwift.xcconfig: Removed.
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift: Renamed from Source/WebCore/WebCoreSwift/CryptoKitShim.swift.
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/crypto/CryptoDigest.h:
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm: Renamed from Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp.
(PAL::toSHA1Context):
(PAL::toSHA224Context):
(PAL::toSHA256Context):
(PAL::toSHA384Context):
(PAL::toSHA512Context):
(PAL::CryptoDigest::CryptoDigest):
(PAL::CryptoDigest::~CryptoDigest):
(PAL::createCryptoDigest):
(PAL::CryptoDigest::create):
(PAL::CryptoDigest::addBytes):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/gcrypt/CryptoDigestGCrypt.cpp:
(PAL::CryptoDigest::CryptoDigest):
(PAL::getGcryptAlgorithm):
(PAL::CryptoDigest::create):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
(PAL::createCryptoDigest):
(PAL::CryptoDigest::create):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::dispatchDigest):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA1.cpp:
(WebCore::CryptoAlgorithmSHA1::digest):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA256.cpp:
(WebCore::CryptoAlgorithmSHA256::digest):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA384.cpp:
(WebCore::CryptoAlgorithmSHA384::digest):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA512.cpp:
(WebCore::CryptoAlgorithmSHA512::digest):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm:

Canonical link: <a href="https://commits.webkit.org/275070@main">https://commits.webkit.org/275070@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef70f213359b07547c05197ab38d0c70e282a35c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40760 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19773 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43138 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43318 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36850 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43067 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22760 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17104 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33802 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41334 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16719 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35131 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14407 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14498 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36120 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44603 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36433 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40186 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15575 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12778 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38530 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17194 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/35402 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9148 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17245 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16838 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/PAL/pal/crypto/CryptoDigest.h:
```diff
@@ -49,6 +49,7 @@ class CryptoDigest {
     PAL_EXPORT void addBytes(const void* input, size_t length);
     PAL_EXPORT Vector<uint8_t> computeHash();
     PAL_EXPORT String toHexString();
+    PAL_EXPORT static std::optional<Vector<uint8_t>> computeHash(Algorithm, Vector<uint8_t>&, bool);
```
const Vector<uint8_t>&

achristensen07 commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm:
```diff
@@ -43,8 +44,7 @@ inline CC_SHA1_CTX* toSHA1Context(CryptoDigestContext* context)
     ASSERT(context->algorithm == CryptoDigest::Algorithm::SHA_224);
-    return static_cast<CC_SHA256_CTX*>(context->ccContext);
-}
+    return static_cast<CC_SHA256_CTX*>(context->ccContext); }
```
} should be on next line.

nmahendru replied:
:/ 
Sorry fixed. Totally unintended.

achristensen07 commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm:
```diff
@@ -88,54 +88,55 @@ CryptoDigest::~CryptoDigest()
 
+std::unique_ptr<CryptoDigest> CryptoDigest::create(CryptoDigest::Algorithm algorithm)
+{
+    std::unique_ptr<CryptoDigest> digest(new CryptoDigest);
```
Use make_unique instead of new.  Otherwise I want to look for a corresponding delete and can't find it.
Same elsewhere in PR.

nmahendru replied:
done.

achristensen07 commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm:
```diff
@@ -88,54 +88,55 @@ CryptoDigest::~CryptoDigest()
 
 
-std::unique_ptr<CryptoDigest> CryptoDigest::create(CryptoDigest::Algorithm algorithm)
+static void* createCryptoDigest(CryptoDigest::Algorithm algorithm)
```
This is a slight improvement, but this is error-prone memory management.  I think we should make CryptoDigestContext a std::variant of 4 different types of unique_ptr, each with their own deleter.  Can be in a separate PR.

nmahendru replied:
Done.

achristensen07 commented on LayoutTests/crypto/subtle/sha1-digest.html:
```diff
@@ -19,6 +19,8 @@
 crypto.subtle.digest("sha-1", message).then(function(result){
     digest = result;
     shouldBe("bytesToHexString(digest)", "expectedDigest");
+    return crypto.subtle.digest("sha-1", new Uint8Array(0));
```
Shouldn't there be an expected digest of an empty array in the test?  I'd expect another shouldBe call.

nmahendru replied:
I have just added the zero length to make sure CryptoKit is not crashing. Although I suspect that this is never used as calculating the digest of a zero length input is not that useful.

achristensen07 reviewed (changes requested):
Looks like the Windows build is broken.

achristensen07 commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -184,3 +187,109 @@ public final class AesKw: NSObject {
+        -> ErrorCodes
+    {
+        if dataSize > Int.max
+            || digestSize == 0
+            || digestSize > Int.max
```
These seem safely impossible.  We don't control dataSize, but we do control these.

nmahendru replied:
they do but then across the boundary all bets are off and I think the performance overhead is probably worth it for future.

achristensen07 commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm:
```diff
@@ -26,116 +26,106 @@
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
     CryptoDigest::Algorithm algorithm;
-    void* ccContext;
+    std::variant<
```
using contextPointer = std::variant<...>;
That way you don't have to write out the variant types multiple times.

achristensen07 commented on Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm:
```diff
@@ -26,116 +26,106 @@
-        CC_SHA1_Init(context);
-ALLOW_DEPRECATED_DECLARATIONS_END
-        return digest;
+        std::unique_ptr<CC_SHA1_CTX> context = WTF::makeUniqueWithoutFastMallocCheck<CC_SHA1_CTX>();
```
auto context = ...
The type is already clearly stated.

nmahendru committed:
Adding SHA* implementations using CryptoKit
rdar://122317816

Reviewed by Alex Christensen.

This PR introduces the CryptoDigest implementations from CryptoKit.

This change also moves the CryptoKitShim into PAL from WebCore to avoid circular dependencies
from WebCore <== PAL. The CryptoDigest Implementation resides in PAL and still needs to use the Shim.
We also add a negative test to for zero length input digests to ensure the new implementation supports it.

Another big change is that the Digest implementations have been replaced with a one shot implementation
instead of init, update and finalize style. That helps simplify the swift shim and does not mandate dealing with swift
allocated memory in C++ land.
The remnants of the init/update/finalize style have been
left in because of this(Just in case we have to implement it in the future):
https://github.com/wintercg/proposal-webcrypto-streams/blob/main/explainer.md

Another change is to introduce smart pointers for managing memory wherever possible.

* LayoutTests/crypto/subtle/sha1-digest.html:
* LayoutTests/crypto/subtle/sha256-digest.html:
* LayoutTests/crypto/subtle/sha384-digest.html:
* LayoutTests/crypto/subtle/sha512-digest.html:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/Configurations/WebCoreSwift.xcconfig: Removed.
* Source/WebCore/PAL/Configurations/PAL.xcconfig:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift: Renamed from Source/WebCore/WebCoreSwift/CryptoKitShim.swift.
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/crypto/CryptoDigest.h:
* Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.mm: Renamed from Source/WebCore/PAL/pal/crypto/commoncrypto/CryptoDigestCommonCrypto.cpp.
(PAL::toSHA1Context):
(PAL::toSHA224Context):
(PAL::toSHA256Context):
(PAL::toSHA384Context):
(PAL::toSHA512Context):
(PAL::CryptoDigest::CryptoDigest):
(PAL::CryptoDigest::~CryptoDigest):
(PAL::createCryptoDigest):
(PAL::CryptoDigest::create):
(PAL::CryptoDigest::addBytes):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/gcrypt/CryptoDigestGCrypt.cpp:
(PAL::CryptoDigest::CryptoDigest):
(PAL::getGcryptAlgorithm):
(PAL::CryptoDigest::create):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/PAL/pal/crypto/openssl/CryptoDigestOpenSSL.cpp:
(PAL::createCryptoDigest):
(PAL::CryptoDigest::create):
(PAL::CryptoDigest::computeHash):
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::dispatchDigest):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA1.cpp:
(WebCore::CryptoAlgorithmSHA1::digest):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA256.cpp:
(WebCore::CryptoAlgorithmSHA256::digest):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA384.cpp:
(WebCore::CryptoAlgorithmSHA384::digest):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmSHA512.cpp:
(WebCore::CryptoAlgorithmSHA512::digest):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm:

Canonical link: https://commits.webkit.org/275070@main
